nomad/node_pool_endpoint.go

```diff
@@ -252,3 +254,92 @@ func (n *NodePool) DeleteNodePools(args *structs.NodePoolDeleteRequest, reply *s
 	reply.Index = index
 	return nil
 }
+
+// Nodes returns the list of nodes that are in the given node pool.
+func (n *NodePool) Nodes(args *structs.NodePoolNodesRequest, reply *structs.NodePoolNodesResponse) error {
+	authErr := n.srv.Authenticate(n.ctx, args)
+	if done, err := n.srv.forward("NodePool.Nodes", args, args, reply); done {
+		return err
+	}
+	n.srv.MeasureRPCRate("node_pool", structs.RateMetricRead, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "node_pool", "nodes"}, time.Now())
+
+	// Resolve ACL token and verify it has read capability for nodes and the
+	// node pool.
+	aclObj, err := n.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+
+	allowed := aclObj.AllowNodeRead() &&
+		aclObj.AllowNodePoolOperation(args.Name, acl.NodePoolCapabilityRead)
+	if !allowed {
+		return structs.ErrPermissionDenied
+	}
+
+	// Ensure node pool name is set.
+	if strings.TrimSpace(args.Name) == "" {
+		return fmt.Errorf("missing node pool name")
+	}
+
+	// Setup blocking query.
+	opts := blockingOptions{
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		run: func(ws memdb.WatchSet, store *state.StateStore) error {
+			// Verify node pool exists.
+			pool, err := store.NodePoolByName(ws, args.Name)
+			if err != nil {
+				return err
+			}
+			if pool == nil {
+				return fmt.Errorf("node pool %q not found", args.Name)
+			}
```

Comment:
Should this be checked in the state store? Other endpoints like this (`/v1/job/:id/allocations` for example) don't even check if `:id` exists 😅

Reply:
Having it here lets us bail out early so we don't have to do all the paginator setup for what's inevitably going to be an empty iterator. We just return `nil` and check again on the next pass if the pool exists, which is particularly nice for a blocking query where the node pool will get created automatically when the first node in the pool registers.

Reply:
Updated to return `nil` instead of error.